### PR TITLE
Parsecontext opm error

### PR DESCRIPTION
**Pull request by joakim-hove**
This is part of the ongoing effort to improve error messages. The main content of this PR is:

1. The `ParseContext::handleError()`will throw `OpmInputError` instead of `std::invalid_argument()`
2. All messages in `OpmInputError` exceptions are formatted consistently.
3. The main parse loop differentiates between `OpmInputError` and `std::exception` - to make sure messages are not repeated.

Downstream: https://github.com/OPM/opm-simulators/pull/2810

### CODE REVIEW

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
@alfbr : This is still work in progress, but the current PR is starting to get ready for some testing. The changes in #1951 have been included.

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
jenkins build this please

File: src/opm/parser/eclipse/EclipseState/Schedule/ArrayDimChecker.cpp [hunk 1]
```diff
@@ -43,13 +46,20 @@ namespace {
...
+                if (location) {
+                    std::string fmt_message = fmt::format("Problem with keyword {{keyword}}\n"
+                                                          "In {{file}} line {{line}}\n"
+                                                          "The case has {0} wells - but only {1} are specified in WELLDIMS.\n"
+                                                          "Please increase item 1 in WELLDIMS to at least {1}", nWells, wdims.maxWellsInField());
+
+                    ctxt.handleError(Opm::ParseContext::RUNSPEC_NUMWELLS_TOO_LARGE,
+                                     fmt_message, *location, guard);
+                } else {
+                    std::string msg = fmt::format("The case does not have a WELDIMS keyword.\n"
```
Comment (bska) [inline src/opm/parser/eclipse/EclipseState/Schedule/ArrayDimChecker.cpp:59]: Tiny nit: The keyword is `WELLDIMS`, with two `L`s.

**Comment by joakim-hove** [comment]
jenkins build this opm-simulators=2810 please

File: src/opm/parser/eclipse/EclipseState/Schedule/ArrayDimChecker.cpp (hunk 1 above)
Comment (joakim-hove) [reply src/opm/parser/eclipse/EclipseState/Schedule/ArrayDimChecker.cpp]: fixed.

**Comment by joakim-hove** [comment]
jenkins build this opm-simulators=2810 please

**Comment by alfbr** [comment]
I love it! This is a big step in the right direction. I believe we really need to replace the number look-up in fmt with strings, or make it very simple to know which number refers to what. Other than that, this PR adresses so many important cases that I believe it should go in now.

**Comment by joakim-hove** [comment]
> I love it! This is a big step in the right direction. 

Good - I am glad to hear!

> I believe we really need to replace the number look-up in fmt with strings,

I thought that was already done; or the only numbered placeholders are evaluated right at site (I really think) - i.e. there is no need to remember/know a numbering convention apart the `fmt::format()` statement you are looking right at. Could of course use named patterns also for those.

File: src/opm/parser/eclipse/EclipseState/Schedule/ArrayDimChecker.cpp [hunk 2]
```diff
@@ -43,13 +46,20 @@ namespace {
...
-                   << "section.  Increase item 1 of WELLDIMS accordingly.";
-
-                ctxt.handleError(Opm::ParseContext::RUNSPEC_NUMWELLS_TOO_LARGE,
-                                 os.str(), guard);
+                const auto& location = wdims.location();
+                if (location) {
+                    std::string fmt_message = fmt::format("Problem with keyword {{keyword}}\n"
+                                                          "In {{file}} line {{line}}\n"
+                                                          "The case has {0} wells - but only {1} are specified in WELLDIMS.\n"
+                                                          "Please increase item 1 in WELLDIMS to at least {1}", nWells, wdims.maxWellsInField());
```
Comment (alfbr) [inline src/opm/parser/eclipse/EclipseState/Schedule/ArrayDimChecker.cpp:54]: Here, I have no idea what {0} and {1} is, but I can guess from the context.

File: src/opm/parser/eclipse/EclipseState/Schedule/ArrayDimChecker.cpp [hunk 3]
```diff
@@ -66,14 +76,20 @@ namespace {
...
-                   << " connections per well in RUNSPEC section.  Increase item "
-                   << "2 of WELLDIMS accordingly.";
-
-                ctxt.handleError(Opm::ParseContext::RUNSPEC_CONNS_PER_WELL_TOO_LARGE,
-                                 os.str(), guard);
+                const auto& location = wdims.location();
+                if (location) {
+                    std::string fmt_message = fmt::format("Problem with keyword {{keyword}}\n"
+                                                          "In {{file}} line {{line}}\n"
+                                                          "The case has a well with {0} connections, but {1} is specified as maxmimum in WELLDIMS.\n"
```
Comment (alfbr) [inline src/opm/parser/eclipse/EclipseState/Schedule/ArrayDimChecker.cpp:83]: Same here

File: src/opm/parser/eclipse/EclipseState/Schedule/ArrayDimChecker.cpp [hunk 4]
```diff
@@ -88,14 +104,21 @@ namespace {
...
-                   << " in RUNSPEC section.  Increase item 3 of WELLDIMS "
-                   << "accordingly.";
-
-                ctxt.handleError(Opm::ParseContext::RUNSPEC_NUMGROUPS_TOO_LARGE,
-                                 os.str(), guard);
+                const auto& location = wdims.location();
+                if (location) {
+                    std::string fmt_message = fmt::format("Problem with keyword {{keyword}}\n"
+                                                          "In {{file}} line {{line}}\n"
+                                                          "The case has {0} non-FIELD groups, the WELLDIMS keyword specifies {1}\n."
```
Comment (alfbr) [inline src/opm/parser/eclipse/EclipseState/Schedule/ArrayDimChecker.cpp:111]: and here

File: src/opm/parser/eclipse/EclipseState/Schedule/ArrayDimChecker.cpp (hunk 2 above)
Comment (joakim-hove) [reply src/opm/parser/eclipse/EclipseState/Schedule/ArrayDimChecker.cpp:54]: OK - it is standard fmt syntax where `{0}` is replaced with the first argument, `{1}` with the second argument and so on; i.e. for this particular example `fmt` will replace
```
{0} -> nWells
{1} -> wdims.maxWellsInField()
```
I can name them.

**Comment by alfbr** [comment]
> > I love it! This is a big step in the right direction.
> 
> Good - I am glad to hear!
> 
> > I believe we really need to replace the number look-up in fmt with strings,
> 
> I thought that was already done; or the only numbered placeholders are evaluated right at site (I really think) - i.e. there is no need to remember/know a numbering convention apart the `fmt::format()` statement you are looking right at. Could of course use named patterns also for those.

Ah, now I see. Sorry for the noise. Actually I did not notice the strings at the end of the line which are human readable. The syntax of fmt is unfamiliar to me. I retract my complaint. From my end this is good.

**Comment by alfbr** [comment]
Sorry, seems you had a comment that got deleted when I removed mine. That was unintentional.

**Comment by joakim-hove** [comment]
> Sorry, seems you had a comment that got deleted when I removed mine. That was unintentional.

No worries

File: src/opm/parser/eclipse/EclipseState/IOConfig/RestartConfig.cpp [hunk 5]
```diff
@@ -332,17 +333,16 @@ inline std::map< std::string, int > RPT( const DeckKeyword& keyword,
 
     if (ints && strs) {
         const auto& location = keyword.location();
-        std::string msg = "Mixed style input is not allowed for keyword: " + keyword.name() + " at " + location.filename + "(" + std::to_string( location.lineno ) + ")";
-        parseContext.handleError(ParseContext::RPT_MIXED_STYLE, msg, errors);
+        std::string msg = "Error in keyword {keyword}, mixing mnemonics and integers is not allowed\n"
```
Comment (atgeirr) [inline src/opm/parser/eclipse/EclipseState/IOConfig/RestartConfig.cpp:336]: I have to say that I find the new style here, with keyword format args, quite readable. I agree with Alf that this looks nice. I think we mostly worked out the disagreements in the discussion about the exception class.

However, you claim that these messages are consistent in some sense. Is that simply by convention? I do not see any common code between these messages? By convention, I mean it looks like many messages end with:
```
... (end-of-line)
"In {file} line {line}."
```

Comment (joakim-hove) [reply src/opm/parser/eclipse/EclipseState/IOConfig/RestartConfig.cpp:336]: > Is that simply by convention? 

yes - it is simply by convention. I agree that maybe that convention should be codified - but I I wanted to get a bit more experience/feel for it before I created a straightjacket for this.

**Comment by joakim-hove** [comment]
jenkins build this opm-simulators=2810 please

**Comment by joakim-hove** [comment]
jenkins build this opm-simulators=2810 please

**Comment by joakim-hove** [comment]
jenkins build this opm-simulators=2810 please

**Comment by joakim-hove** [comment]
jenkins build this opm-simulators=2810 please

**Comment by joakim-hove** [comment]
jenkins build this opm-simulators=2810 please

**Comment by joakim-hove** [comment]
jenkins build this opm-simulators=2810 please

File: src/opm/parser/eclipse/Parser/ParseContext.cpp [hunk 6]
```diff
@@ -158,7 +162,7 @@ namespace Opm {
             // make sure the error object does not terminate the application
             // when it goes out of scope.
             errors.clear();
-            throw std::invalid_argument(errorKey + ": " + msg);
+            throw OpmInputError(msg_fmt, *location);
```
Comment (akva2) [inline src/opm/parser/eclipse/Parser/ParseContext.cpp:165]: unconditional deref of an optional. this causes the errors on jenkins.

File: src/opm/parser/eclipse/EclipseState/SummaryConfig/SummaryConfig.cpp [hunk 7]
```diff
@@ -565,8 +560,9 @@ inline void keywordR2R( SummaryConfig::keyword_list& /* list */,
                         const DeckKeyword& keyword)
 {
     const auto& location = keyword.location();
-    std::string msg = "Region to region summary keyword: " + keyword.name() + " at " + location.filename + ", line " + std::to_string(location.lineno) + " is ignored";
-    parseContext.handleError(ParseContext::SUMMARY_UNHANDLED_KEYWORD, msg, errors);
+    std::string msg_fmt = "Region to region summary keyword {keyword} is ignored\n";
```
Comment (akva2) [inline src/opm/parser/eclipse/EclipseState/SummaryConfig/SummaryConfig.cpp:563]: unintended ;

**Comment by joakim-hove** [comment]
Debugging help from @akva2  and @bska is acknowledged - hopefully it comes through green now!

**Comment by joakim-hove** [comment]
jenkins build this opm-simulators=2810 please

**Pull request merged**